I decided to treat this as a classification problem by creating a new binary variable affair
(did the woman have at least one affair?) and trying to predict the classification for each
woman.
Dataset
The dataset I chose is the affairs dataset that comes with Statsmodels. It was derived
from a survey of women in 1974 by Redbook magazine, in which married women were
asked about their participation in extramarital affairs. More information about the study
is available in a 1978 paper from the Journal of Political Economy.
Description of Variables
The dataset contains 6366 observations of 9 variables:
rate_marriage: woman's rating of her marriage (1 = very poor, 5 = very good)
age: woman's age
yrs_married: number of years married
children: number of children
religious: woman's rating of how religious she is (1 = not religious, 4 = strongly religious)
educ: level of education (9 = grade school, 12 = high school, 14 = some college, 16 =
college graduate, 17 = some graduate school, 20 = advanced degree)
occupation: woman's occupation (1 = student, 2 = farming/semi-skilled/unskilled, 3 =
"white collar", 4 = teacher/nurse/writer/technician/skilled, 5 = managerial/business, 6 =
professional with advanced degree)

occupation_husb: husband's occupation (same coding as above)
affairs: time spent in extra-marital affairs

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
from patsy import dmatrices
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn import metrics
from sklearn.cross_validation import cross_val_score
dta = sm.datasets.fair.load_pandas().data

C:\Users\pc\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
dta.head()

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2            4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3            4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4            5.0  27.0          9.0       1.0        1.0  14.0         3.0   

   occupation_husb   affairs  
0              5.0  0.111111  
1              4.0  3.230769  
2              5.0  1.400000  
3              5.0  0.727273  
4              4.0  4.666666

In [3]:
# add "affair" column: 1 represents having affairs, 0 represents not
dta['affair'] = (dta.affairs > 0).astype(int)
dta.head(2)

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   

   occupation_husb   affairs  affair  
0              5.0  0.111111       1  
1              4.0  3.230769       1

In [5]:
y, X = dmatrices('affair ~ rate_marriage + age + yrs_married + children + religious + educ + C(occupation) + C(occupation_husb)',
dta, return_type="dataframe")
X.head(2)

Intercept  C(occupation)[T.2.0]  C(occupation)[T.3.0]  \
0        1.0                   1.0                   0.0   
1        1.0                   0.0                   1.0   

   C(occupation)[T.4.0]  C(occupation)[T.5.0]  C(occupation)[T.6.0]  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   

   C(occupation_husb)[T.2.0]  C(occupation_husb)[T.3.0]  \
0                        0.0                        0.0   
1                        0.0                        0.0   

   C(occupation_husb)[T.4.0]  C(occupation_husb)[T.5.0]  \
0                        0.0                        1.0   
1                        1.0                        0.0   

   C(occupation_husb)[T.6.0]  rate_marriage   age  yrs_married  children  \
0                        0.0            3.0  32.0          9.0       3.0   
1                        0.0            3.0  27.0         13.0       3.0   

   religious  educ  
0        3.0  17.0  
1        1.0  14.0

In [19]:
#To give columns some better and readable name

X = X.rename(columns = {'C(occupation)[T.2.0]':'occ_2',
'C(occupation)[T.3.0]':'occ_3',
'C(occupation)[T.4.0]':'occ_4',
'C(occupation)[T.5.0]':'occ_5',
'C(occupation)[T.6.0]':'occ_6',
'C(occupation_husb)[T.2.0]':'occ_husb_2',
'C(occupation_husb)[T.3.0]':'occ_husb_3',
'C(occupation_husb)[T.4.0]':'occ_husb_4',
'C(occupation_husb)[T.5.0]':'occ_husb_5',
'C(occupation_husb)[T.6.0]':'occ_husb_6'})

X.head()

Intercept  occ_2  occ_3  occ_4  occ_5  occ_6  occ_husb_2  occ_husb_3  \
0        1.0    1.0    0.0    0.0    0.0    0.0         0.0         0.0   
1        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
2        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
3        1.0    0.0    0.0    0.0    1.0    0.0         0.0         0.0   
4        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   

   occ_husb_4  occ_husb_5  occ_husb_6  rate_marriage   age  yrs_married  \
0         0.0         1.0         0.0            3.0  32.0          9.0   
1         1.0         0.0         0.0            3.0  27.0         13.0   
2         0.0         1.0         0.0            4.0  22.0          2.5   
3         0.0         1.0         0.0            4.0  37.0         16.5   
4         1.0         0.0         0.0            5.0  27.0          9.0   

   children  religious  educ  
0       3.0        3.0  17.0  
1       3.0        1.0  14.0  
2       0.0        1.0  16.0  
3       4.0        3.0  16.0  
4       1.0        1.0  14.0

In [8]:
y = np.ravel(y)
y

array([1., 1., 1., ..., 0., 0., 0.])

In [ ]:
#Now we run logistic regression on the entire data set after fitting the X and y values

In [9]:
model=LogisticRegression()
model=model.fit(X,y)

In [10]:
model.score(X,y)

0.7258875274897895

In [11]:
y.mean()

0.3224945020420987

This implies that 32.2% of women had an affair.

In [20]:
pd.DataFrame(list(zip(X.columns,np.transpose(model.coef_))))

0                       1
0       Intercept     [1.489835891324933]
1           occ_2   [0.18806639024440983]
2           occ_3    [0.4989478668156914]
3           occ_4   [0.25066856498524825]
4           occ_5    [0.8390080648117001]
5           occ_6    [0.8339084337443315]
6      occ_husb_2    [0.1906359445867889]
7      occ_husb_3    [0.2978327129263421]
8      occ_husb_4    [0.1614088540760616]
9      occ_husb_5   [0.18777091388972483]
10     occ_husb_6   [0.19401637225511495]
11  rate_marriage   [-0.7031233597323255]
12            age  [-0.05841777448168919]
13    yrs_married   [0.10567653799735635]
14       children  [0.016919266970905608]
15      religious   [-0.3711362653137546]
16           educ   [0.00401650319563816]

Above data shows 3 negative coefficients, out of which marriage rating and religious aresignificantly -ve. This implies that those who are more religious or have ratedtheir marriage high have less chances of an affair asthey are inversely related.

In [24]:
#Now we will create a train test split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)
model2=LogisticRegression()
model2.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [25]:
# Now we will predict the outcome for the test set
predicted=model2.predict(X_test)
predicted

array([1., 0., 0., ..., 0., 0., 0.])

In [30]:
# Now we will check the accuracy score on the test set
metrics.accuracy_score(y_test,predicted)

0.7298429319371728

In [ ]:
# the accuracy comes similar to that werecievedabove with our train data

Now checking this model foran unknown example like on a woman 41-year-old house wife who has an advance degree,, has been married for 12 years, has 2 child, rates herself as strongly religious, rates her marriage as very good, and her husband is a professional with an advance degree.

In [32]:
model.predict_proba(np.array([[1,0,0,1,0,0,0,0,0,0,1,5,42,12,2,5,20]]))

array([[0.95338964, 0.04661036]])

In [ ]:
Hence predicted that the prebability ofan affairis 4.6%